src/fsharp/pickle.fs

```diff
@@ -1411,15 +1411,41 @@ let u_trait st =
 #if INCLUDE_METADATA_WRITER
 let p_rational q st = p_int32 (GetNumerator q) st; p_int32 (GetDenominator q) st
 
-let rec p_measure_expr unt st =
-    let unt = stripUnitEqnsAux false unt 
-    match unt with 
-    | MeasureCon tcref   -> p_byte 0 st; p_tcref "measure" tcref st
-    | MeasureInv x       -> p_byte 1 st; p_measure_expr x st
-    | MeasureProd(x1,x2) -> p_byte 2 st; p_measure_expr x1 st; p_measure_expr x2 st
-    | MeasureVar(v)      -> p_byte 3 st; p_tpref v st
-    | MeasureOne         -> p_byte 4 st
-    | MeasureRationalPower(x,q) -> p_byte 5 st; p_measure_expr x st; p_rational q st
+let p_measurecon tcref st = p_byte 0 st; p_tcref "measure" tcref st
+let p_measurevar v st = p_byte 3 st; p_tpref v st
+
+let p_measure_varcon unt st =
+     match unt with 
+     | MeasureCon tcref   -> p_measurecon tcref st
+     | MeasureVar v       -> p_measurevar v st
+     | _                  -> pfailwith st ("p_measure_varcon: expected measure variable or constructor")
+
+let rec p_measure_intpower unt n st =
+  match n with
+  | 0 -> p_byte 4 st
```

Comment:
Duplicating "p_byte 4" is subtle. Factor out a new function called "p_measure_zero". And add a comment that "pbyte 4" is the F# canoncical encoding for MeasureZero.